### PR TITLE
[GPU/OpenCL] OpenCL pipeline for layer execution on GPU @open sesame 03/07 10:42

### PR DESCRIPTION
**Pull request by s-debadri**
This PR includes the initial version of OpenCL pipeline for layers to run on GPU. Following wrappers have been added which uses OpenCL APIs internally:

- `opencl_loader`: Loading required OpenCL functions.
- `opencl_context_manager`: Getting device details and managing context creation globally.
- `opencl_command_queue_manager`: Creation and deletion of global command queue, reading and writing buffers, dispatching kernels.
- `opencl_kernel`: Managing kernels and setting arguments.
- `opencl_program`: Building and managing OpenCL program.
- `opencl_buffer`: Creating, reading and writing data on OpenCL buffer.

`opencl_op_interface` has been created to handle the workflow of kernel execution.

Tensor level changes:

- Added `cl_interface` to handle tensor operations which will be executed on GPU.
- Added experimental naïve OpenCL kernel for SGEMV which can be useful for testing the initial pipeline.
- Modified `sum` and `sum_by_batch` function signatures to use boolean flag for choosing compute engine (CPU/GPU).

`enable-opencl` flag incorporated in `meson.options`

**Update (23 Feb 2024)**:

- `ml::train::LayerComputeEngine` enum added to handle compute engine information.  Only `FullyConnected` layer API signature modified as of now.
- `createLayer ` API modified to propagate compute engine info via `factory` to `layer_node`. 
- `cl_context` created to handle global configuration of OpenCL environment which will create command queue and context for OpenCL and register layers to be run on GPU.
- Getter and setter for compute engine added to `layer_context` along with utility to create kernel.
- Modified build configurations as required.
- Added UT for `cl_context` global instance creation with GPU command queue.

> `opencl_op_interface` can be deprecated later. Restored `sum` and `sum_by_batch` function signatures as tensor level kernels are not required at this moment. Refactored experimental naïve OpenCL kernel for SGEMV.

Signed-off-by: Debadri Samaddar <s.debadri@samsung.com>

### CODE REVIEW

File: nntrainer/opencl/meson.build [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  'opencl_op_interface.hpp'
+]
+
+foreach s : opencl_sources
+  nntrainer_sources += meson.current_source_dir() / s
+endforeach
+
+foreach h : opencl_headers
+  nntrainer_headers += meson.current_source_dir() / h
+endforeach
```
Comment (jijoongmoon) [inline nntrainer/opencl/meson.build:22]: please add new line at the end of the file.

File: nntrainer/meson.build [hunk 2]
```diff
@@ -42,7 +42,8 @@ nntrainer_elements = [
   'optimizers',
   'tensor',
   'utils',
-  'graph'
+  'graph',
+  'opencl'
```
Comment (jijoongmoon) [inline nntrainer/meson.build:46]: This can be optional. how about adding ```option('enable-opencl', type: 'boolean', value: false)``` in meson_options.txt

Comment (s-debadri) [reply nntrainer/meson.build]: Added related changes in next commit. Updated description.

File: nntrainer/opencl/meson.build (hunk 1 above)
Comment (s-debadri) [reply nntrainer/opencl/meson.build]: Resolved in next commit

File: nntrainer/opencl/meson.build [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  'third_party/cl_platform.h',
+  'opencl_op_interface.hpp'
+]
+
+foreach s : opencl_sources
+  nntrainer_sources += meson.current_source_dir() / s
+endforeach
+
+foreach h : opencl_headers
+  nntrainer_headers += meson.current_source_dir() / h
```
Comment (jijoongmoon) [inline nntrainer/opencl/meson.build:21]: It might be better to use ```.h``` for header files.

File: nntrainer/opencl/opencl_buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ */
+
+#include "opencl_buffer.hpp"
+
+#include "opencl_loader.hpp"
+
+#include <nntrainer_log.h>
+
+namespace nntrainer::internal {
```
Comment (jijoongmoon) [inline nntrainer/opencl/opencl_buffer.cpp:20]: How about ```c++ nntrainer::opencl ``` namesapce?

File: nntrainer/opencl/opencl_buffer.hpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+#ifndef GPU_CL_OPENCL_BUFFER_HPP_
+#define GPU_CL_OPENCL_BUFFER_HPP_
+
+#include "opencl_command_queue_manager.hpp"
+#include "opencl_context_manager.hpp"
+#include "third_party/cl.h"
+
+namespace nntrainer::internal {
```
Comment (jijoongmoon) [inline nntrainer/opencl/opencl_buffer.hpp:21]: How about adding more comments following the oxygen to provide a better understanding?

File: nntrainer/opencl/opencl_context_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ * @bug     No known bugs except for NYI items
+ * @brief   OpenCL wrapper for context management
+ *
+ */
+
+#include "opencl_context_manager.hpp"
+
+#include <vector>
+
+#include "opencl_loader.hpp"
```
Comment (jihochu) [inline nntrainer/opencl/opencl_context_manager.cpp:18]: please check order.

File: nntrainer/opencl/opencl_kernel.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace nntrainer::internal {
+
+bool Kernel::CreateKernelFromProgram(Program program,
+                                     const std::string &function_name) {
+  int error_code;
+  cl_program prgm = program.GetProgram();
+  kernel_ = clCreateKernel(prgm, function_name.c_str(), &error_code);
+  if (!kernel_ || error_code != CL_SUCCESS) {
+    kernel_ = nullptr;
```
Comment (jihochu) [inline nntrainer/opencl/opencl_kernel.cpp:28]: As a doc, clCreateKernel returns null with error code whenever it fails.

**Comment by jijoongmoon** [comment]
please add the doxygen comment in the codes.

**Review by jijoongmoon — APPROVED** [review]
Great Work! LGTM. 
please Check the CI (oxygen tag and cppcheck)

File: nntrainer/cl_context.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+std::mutex cl_factory_mutex;
+
+std::once_flag global_cl_context_init_flag;
+
+static void add_default_object(ClContext &cc) {
+  using LayerType = ml::train::LayerType;
+
+  cc.registerFactory(nntrainer::createLayer<FullyConnectedLayer>,
+                     FullyConnectedLayer::type, LayerType::LAYER_FC);
```
Comment (djeong20) [inline nntrainer/cl_context.cpp:28]: seems `LayerType` is only used below. could be removed.

```suggestion
  cc.registerFactory(nntrainer::createLayer<FullyConnectedLayer>,
                     FullyConnectedLayer::type, ml::train::LayerType::LAYER_FC);
```

File: nntrainer/opencl/opencl_buffer.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * @brief Copy constructor for buffer
+ *
+ * @param buffer
+ */
+Buffer::Buffer(Buffer &&buffer) :
+  mem_buf_(buffer.mem_buf_),
+  size_(buffer.size_) {
+  buffer.mem_buf_ = nullptr;
+  buffer.size_ = 0;
```
Comment (djeong20) [inline nntrainer/opencl/opencl_buffer.cpp:59]: could you explain why resetting the input buffer?

File: nntrainer/tensor/cl_operations/cl_sgemv.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+ * @author  Debadri Samaddar <s.debadri@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   Experimental SGEMV implementation using OpenCL
+ *
+ * @note This file is experimental and is kept for testing purpose
+ *
+ */
+
+#include "cl_sgemv.h"
+#include <iostream>
```
Comment (djeong20) [inline nntrainer/tensor/cl_operations/cl_sgemv.cpp:17]:
```suggestion
```

File: nntrainer/tensor/cl_operations/cl_sgemv.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+#include <nntrainer_log.h>
+
+namespace nntrainer::internal {
+
+template <typename T>
+T *GpuCLSgemv::CLSgemv(const T *matAdata, const T *vecXdata, T *vecYdata,
+                       T alpha, T beta, unsigned int dim1, unsigned int dim2) {
+
+  ml_logi("GpuCLSgemv::CLSgemv");
```
Comment (djeong20) [inline nntrainer/tensor/cl_operations/cl_sgemv.cpp:28]: is this needed?

File: nntrainer/opencl/opencl_buffer.cpp (hunk 9 above)
Comment (s-debadri) [reply nntrainer/opencl/opencl_buffer.cpp:59]: This is a move constructor. To copy over the data and delete from the previous. Updating the doc in next commit.

File: nntrainer/tensor/cl_operations/cl_sgemv.cpp (hunk 11 above)
Comment (s-debadri) [reply nntrainer/tensor/cl_operations/cl_sgemv.cpp:28]: This file has been kept for testing/experimental purpose as of now, hence the logs. Might be removed later.

Comment (djeong20) [reply nntrainer/tensor/cl_operations/cl_sgemv.cpp:27]: gotcha

File: nntrainer/cl_context.cpp (hunk 8 above)
Comment (s-debadri) [reply nntrainer/cl_context.cpp]: Modified in latest commit

File: nntrainer/tensor/cl_operations/cl_sgemv.h [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    })";
+
+public:
+  template <typename T>
+  T *CLSgemv(const T *matAdata, const T *vecXdata, T *vecYdata, T alpha, T beta,
+             unsigned int dim1, unsigned int dim2);
+};
+} // namespace nntrainer::internal
+
+#endif // GPU_CL_SGEMV_HPP_
```
Comment (djeong20) [inline nntrainer/tensor/cl_operations/cl_sgemv.h:40]:
```suggestion
#endif // __CL_SGEMV_H__
```

File: nntrainer/tensor/cl_operations/cl_sgemv.h [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+ * @author  Debadri Samaddar <s.debadri@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   Experimental SGEMV implementation using OpenCL
+ *
+ * @note This file is experimental and is kept for testing purpose
+ *
+ */
+
+#ifndef GPU_CL_SGEMV_HPP_
+#define GPU_CL_SGEMV_HPP_
```
[the remark this answers is not in the document]
Comment (s-debadri) [reply nntrainer/tensor/cl_operations/cl_sgemv.h]: Resolved in latest commit.

File: nntrainer/tensor/cl_operations/cl_sgemv.h (hunk 12 above)
Comment (s-debadri) [reply nntrainer/tensor/cl_operations/cl_sgemv.h]: Resolved in latest commit.

**Review by DonghakPark — APPROVED** [review]
Hi you got 3 formatting error 

clang-format reports: 3 file(s) not formatted
nntrainer/opencl/opencl_buffer.cpp
nntrainer/opencl/third_party/cl.h
nntrainer/opencl/third_party/cl_platform.h

on here, now our TAOS CI and gitaction CI has different version 
so please fomatting based on gitaction CI ?

**Comment by DonghakPark** [comment]
@s-debadri 
in case of third party code you can ignore doxygen ci error
And if you want to fix, cpp or clang ci error --> you can see details --> git action summary or cmd log

**Comment by skykongkong8** [comment]
I found this commit emits error when using ndk-build in the current main
```bash
ld: error: undefined symbol: nntrainer::ClContext::Global()
>>> referenced by layers_dependent_common_tests.cpp:38 (../unittest/layers/layers_dependent_common_tests.cpp:38)
>>>               /home/sungsik/nntrainer/test/obj/local/arm64-v8a/objs/unittest_layers/__/unittest/layers/layers_dependent_common_tests.o:(LayerSemantics_createFromClContext_pn_Test::TestBody())

ld: error: undefined symbol: int const nntrainer::ClContext::registerFactory<nntrainer::Layer>(std::__ndk1::function<std::__ndk1::unique_ptr<nntrainer::Layer, std::__ndk1::default_delete<nntrainer::Layer> > (std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > const&)>, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, int)
>>> referenced by layers_dependent_common_tests.cpp:40 (../unittest/layers/layers_dependent_common_tests.cpp:40)
>>>               /home/sungsik/nntrainer/test/obj/local/arm64-v8a/objs/unittest_layers/__/unittest/layers/layers_dependent_common_tests.o:(LayerSemantics_createFromClContext_pn_Test::TestBody())

ld: error: undefined symbol: nntrainer::opencl::ContextManager::ReleaseContext()
>>> referenced by layer_context.h:829 (/home/sungsik/nntrainer/nntrainer/layers/layer_context.h:829)
>>>               /home/sungsik/nntrainer/test/obj/local/arm64-v8a/objs/unittest_layers/__/unittest/layers/layers_golden_tests.o:(nntrainer::RunLayerContext::~RunLayerContext())
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by s-debadri** [comment]
> I found this commit emits error when using ndk-build in the current main
> 
> ```shell
> ld: error: undefined symbol: nntrainer::ClContext::Global()
> >>> referenced by layers_dependent_common_tests.cpp:38 (../unittest/layers/layers_dependent_common_tests.cpp:38)
> >>>               /home/sungsik/nntrainer/test/obj/local/arm64-v8a/objs/unittest_layers/__/unittest/layers/layers_dependent_common_tests.o:(LayerSemantics_createFromClContext_pn_Test::TestBody())
> 
> ld: error: undefined symbol: int const nntrainer::ClContext::registerFactory<nntrainer::Layer>(std::__ndk1::function<std::__ndk1::unique_ptr<nntrainer::Layer, std::__ndk1::default_delete<nntrainer::Layer> > (std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > const&)>, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, int)
> >>> referenced by layers_dependent_common_tests.cpp:40 (../unittest/layers/layers_dependent_common_tests.cpp:40)
> >>>               /home/sungsik/nntrainer/test/obj/local/arm64-v8a/objs/unittest_layers/__/unittest/layers/layers_dependent_common_tests.o:(LayerSemantics_createFromClContext_pn_Test::TestBody())
> 
> ld: error: undefined symbol: nntrainer::opencl::ContextManager::ReleaseContext()
> >>> referenced by layer_context.h:829 (/home/sungsik/nntrainer/nntrainer/layers/layer_context.h:829)
> >>>               /home/sungsik/nntrainer/test/obj/local/arm64-v8a/objs/unittest_layers/__/unittest/layers/layers_golden_tests.o:(nntrainer::RunLayerContext::~RunLayerContext())
> clang++: error: linker command failed with exit code 1 (use -v to see invocation)
> ```

@skykongkong8 
This requires `ENABLE_OPENCL` to be `true` in `meson_options.txt` while building `libnntrainer.so`.
Or, `nntrainer/test/jni/Android.mk` can be modified as such:
- Modify setup for `unittest_layers` and remove `-DENABLE_OPENCL=1` from `LOCAL_CFLAGS` variable.
- This will disable OpenCL dependency for `unittest_layers`.

**Comment by skykongkong8** [comment]
@s-debadri
Right.. then could you make a PR for that issue?

**Pull request merged**